pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,264 @@
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gardener/gardener/pkg/nodeagent/dbus"
+)
+
+const (
+	// kubeletServiceName is the systemd service name of the kubelet.
+	kubeletServiceName = "kubelet.service"
+	// defaultKubeletHealthEndpoint is the health endpoint of the kubelet.
+	defaultKubeletHealthEndpoint = "http://127.0.0.1:10248/healthz"
+	// maxToggles defines how often the kubelet can change the readiness during toggleTimeSpan until the node will be rebooted.
+	maxToggles = 5
+	// toggleTimeSpan is a floating time window where the kubelet readiness toggles are considered harmful.
+	toggleTimeSpan = 10 * time.Minute
+)
+
+// KubeletHealthChecker configures the kubelet healthcheck.
+type KubeletHealthChecker struct {
+	// Clock exported for testing.
+	Clock clock.Clock
+	// KubeletReadinessToggles contains an entry for every toggle between kubelet Ready->NotReady or NotReady->Ready state.
+	KubeletReadinessToggles []time.Time
+	// NodeReady indicates if the node is ready. Exported for testing.
+	NodeReady bool
+
+	client client.Client
+	// firstFailure stores the time of the first failed kubelet health check
+	firstFailure *time.Time
+	dbus         dbus.DBus
+	recorder     record.EventRecorder
+	// lastInternalIP stores the node internalIP
+	lastInternalIP netip.Addr
+	// getAddresses is a func which returns a slice of ip addresses on this node
+	getAddresses          func() ([]net.Addr, error)
+	kubeletHealthEndpoint string
+}
+
+// NewKubeletHealthChecker create a instance of a kubelet health check.
+func NewKubeletHealthChecker(client client.Client, clock clock.Clock, dbus dbus.DBus, recorder record.EventRecorder, getAddresses func() ([]net.Addr, error)) *KubeletHealthChecker {
+	return &KubeletHealthChecker{
+		client:                  client,
+		dbus:                    dbus,
+		Clock:                   clock,
+		recorder:                recorder,
+		getAddresses:            getAddresses,
+		KubeletReadinessToggles: []time.Time{},
+		kubeletHealthEndpoint:   defaultKubeletHealthEndpoint,
+	}
+}
+
+// Name returns the name of this health check.
+func (*KubeletHealthChecker) Name() string {
+	return "kubelet"
+}
+
+// HasLastInternalIP returns true if the node.InternalIP was stored.
+// Exported for testing.
+func (k *KubeletHealthChecker) HasLastInternalIP() bool {
+	return k.lastInternalIP.IsValid()
+}
+
+// SetKubeletHealthEndpoint set the kubeletHealthEndpoint.
+// Exported for testing.
+func (k *KubeletHealthChecker) SetKubeletHealthEndpoint(kubeletHealthEndpoint string) {
+	k.kubeletHealthEndpoint = kubeletHealthEndpoint
+}
+
+// Check performs the actual health check for the kubelet.
+func (k *KubeletHealthChecker) Check(ctx context.Context, node *corev1.Node) error {
+	log := logf.FromContext(ctx).WithName("kubelet")
+
+	// This mimics the old behavior of the kubelet-health-monitor which checks for
+	// to many NotReady->Ready toggles in a certain time and reboots the node in such a case.
```

Comment:
Should we not rather describe what our intend is here instead of referring to the "old behavior"? Soon, the old behavior will be deleted. Then the comment might be confusing.